### ISSUE TITLE
RFC: Deprecate cmd/integration/integration.go

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#24440]
The integration test is hard to debug -- it runs tests concurrently with interspersed logs. I doubt that it's catching any bug that neither unit test nor a e2e test can catch, and both unit and e2e tests are easier to debug.
I'd vote for deprecating this file and remove/move all the tests. Any objections?

Here is a list of tests that it curretnly runs: 
https://github.com/kubernetes/kubernetes/blob/f0c33d65b6fea84d791059d96c93a15e4b3de4ec/cmd/integration/integration.go#L1022
```
		runReplicationControllerTest,
		runAtomicPutTest,
		runPatchTest,
		runServiceTest,
		runAPIVersionsTest,
		runMasterServiceTest,
		func(c *client.Client) {
			runSelfLinkTestOnNamespace(c, api.NamespaceDefault)
			runSelfLinkTestOnNamespace(c, "other")
		},
```

**Comment by bprashanth** [comment]
Yeah the rc and services tests are redundant, I think we exercise them enough in e2es.

Don't know about: https://github.com/kubernetes/kubernetes/blob/f0c33d65b6fea84d791059d96c93a15e4b3de4ec/cmd/integration/integration.go#L1086
It quicky creates/deletes pods with the same hostport that get scheduled onto a single node. Don't see a reason why we can't move it to an e2e, but I don't think it's been done yet.

**Comment by yujuhong** [comment]
@lavalamp, since you are the original author of `runSchedulerNoPhantomPodsTest`, do you have any suggestions/preferences about this test?

**Comment by lavalamp** [comment]
That test is important, it has caught regressions two or three times. It
should be movable to the other integration test section.

**Comment by mml** [comment]
@yujuhong looks like you're working on this already.  Feel free to unassign if that's incorrect.

**Comment by yujuhong** [comment]
I uploaded a PR to remove rc and service tests based on https://github.com/kubernetes/kubernetes/issues/24440#issuecomment-212229355

I'll let each team decide whether they want to migrate or delete the [tests](https://github.com/kubernetes/kubernetes/blob/f0c33d65b6fea84d791059d96c93a15e4b3de4ec/cmd/integration/integration.go#L1022).

**Comment by bprashanth** [comment]
@lavalamp @kubernetes/sig-scheduling maybe someone with spare time can rewrite that scheduler test as an e2e, and we can kill integration? 

I continue to see integration flake that's hard to debug, current SQ closing failure: 
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/25446/kubernetes-pull-test-unit-integration/26688/

Why did that fail? Ctrl+f "fail" gives me 700 matches, scrolling through the logs i see all sorts of nonsense. 
```
        "reason": "KubeletNotReady",
        "message": "Failed to start ContainerManager system validation failed - Following Cgroup subsystem not mounted: [memory]"
...


E0519 07:55:54.283115   20988 etcd_watcher.go:448] failure to decode api object: empty data
from &client.Response{Action:"delete", Node:(*client.Node)(0xc82d58e360), PrevNode:(*client.Node)(0xc82d58e3c0), Index:0x316} &client.Node{Key:"/registry", Dir:true, Value:"", Nodes:client.Nodes(nil), CreatedIndex:0x305, ModifiedIndex:0x317, Expiration:(*time.Time)(nil), TTL:0}
E0519 07:55:54.283149   20988 etcd_watcher.go:448] failure to decode api object: emp

...

fit failure on node (node-multi-scheduler-test-node): Insufficient CPU
; retrying
E0519 07:57:24.818591   20988 factory.go:481] Error scheduling default pod-test-allocatable2: pod (pod-test-allocatable2) failed to fit in any node
fit failure on node (node-allocatable-scheduler-test-node): Insufficient CPU
; retrying
```

I'm going to guess that scheduler test wasn't able to put pods on nodes because of the kubeletNotReady condition, but that's just a guess.

**Comment by hongchaodeng** [comment]
Those failures even existed in passing integration test as well (I can feel your pain).

> Insufficient CPU 

This is due to [TestUnschedulableNodes()](https://github.com/kubernetes/kubernetes/blob/8a04506546db27bab6b3a58ca88378fadfdb0541/test/integration/scheduler_test.go).

It would be great if original author of those tests can clarify what each test does and maintain it. Otw those tests are just legacy obstacles that we should get rid of.

**Comment by bprashanth** [comment]
the rc and services tests have been removed. The only ones remaining are put/patch/apiserver typs tests and the scheduler phantom pod tests IIRC. The former is relatively straightforward and only needs to apiserver, the latter would be much easier to maintain as an e2e. Bandwidth issue.

**Comment by hongchaodeng** [comment]
> scheduler phantom pod tests ... latter would be much easier to maintain as an e2e

I see. I can definitely help with it. Let me spend some time figuring them out. Meanwhile, anyone familiar with the scheduler tests should feel free to open another issue and track them.

**Comment by hongchaodeng** [comment]
@davidopp 
What's your opinion on this?

**Comment by thockin** [comment]
I agree that integration either needs to be killed or massively overhauled

**Comment by bprashanth** [comment]
This issue describes the circumstances leading up to that scheduler integration test, for anyone attempting to port it: https://github.com/kubernetes/kubernetes/issues/6023. Though IIRC we got rid of the modeler, so things might be different now.

**Comment by bprashanth** [comment]
I think we actually have a spike in integration test failures: https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-test-go/12298

**Comment by yujuhong** [comment]
> I think we actually have a spike in integration test failures: https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-test-go/12298

@bprashanth unfortunately the failure you linked was not from this file (`cmd/integration/integration.go`)

```
TestCascadingDeletion

garbage_collector_test.go:211: timed out waiting for the condition
```
Those tests were in `test/integration/`, and were considered easier to debug (at least from @lavalamp's doc last time). Personally I don't have much experiences debugging this new category of integration tests, so I can't say much. They do seem to flake a lot recently in per-PR tests as well.

**Comment by hongchaodeng** [comment]
> Those tests were in test/integration/

I concur on this regarding the "fail" message seen on scheduler test too.

Re. "SchedulerNoPhantomPodsTest", it assumes two kubelets with predefined resources, which is easier to maintain in simulated environment. Right? If testing in integration.go is to hard to debug, we could make it a lightweight unit test. IIRC, it's testing one specific race condition that existed before in scheduler.

**Comment by bprashanth** [comment]
oh, hmm, different bug then https://github.com/kubernetes/kubernetes/issues/25890

Re shceduler test: If you can manage with a unittest that would be great. If not I'm fine running the integration tests serially. I don't think it'll take that much longer, and flakes that only show up in parallel runs are hopefully covered in gce-e2e.
